File: frontend/api/attendance.ts
import { API_BASE_URL, API_ENDPOINTS, getAuthHeaders } from './config';
import { AttendanceRecord } from '../types';

export interface AttendanceAPIRecord {
  id: number;
  user_id: number;
  date: string;
  start_time: string;
  end_time: string;
  break_minutes: number;
  report: string;
  created_at: string;
  updated_at: string;
}

export interface CreateAttendanceDTO {
  date: string;
  start_time: string;
  end_time: string;
  break_minutes: number;
  report: string;
}

export const attendanceAPI = {
  async getMyAttendances(month?: string): Promise<AttendanceAPIRecord[]> {
    const params = month ? `?month=${month}` : '';
    const response = await fetch(`${API_BASE_URL}${API_ENDPOINTS.ATTENDANCE}${params}`, {
      headers: getAuthHeaders(),
    });

    if (!response.ok) {
      throw new Error('Failed to fetch attendances');
    }

    return response.json();
  },

  async createAttendance(data: CreateAttendanceDTO): Promise<AttendanceAPIRecord> {
    const response = await fetch(`${API_BASE_URL}${API_ENDPOINTS.ATTENDANCE}`, {
      method: 'POST',
      headers: getAuthHeaders(),
      body: JSON.stringify(data),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.error || 'Failed to create attendance');
    }

    return response.json();
  },

  async updateAttendance(id: number, data: CreateAttendanceDTO): Promise<AttendanceAPIRecord> {
    const response = await fetch(`${API_BASE_URL}${API_ENDPOINTS.ATTENDANCE_BY_ID(id)}`, {
      method: 'PUT',
      headers: getAuthHeaders(),
      body: JSON.stringify(data),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.error || 'Failed to update attendance');
    }

    return response.json();
  },
};

File: backend/internal/domain/entity/attendance.go
package entity

import (
	"errors"
	"time"
)

type Attendance struct {
	Id           int
	UserId       int
	Date         time.Time
	StartTime    time.Time
	EndTime      time.Time
	BreakMinutes int
	Report       string
	CreatedAt    time.Time
	UpdatedAt    time.Time
}

func NewAttendance(userId int, date, startTime, endTime time.Time, breakMinutes int, report string) (*Attendance, error) {
	if userId <= 0 {
		return nil, errors.New("invalid user ID")
	}
	if startTime.After(endTime) {
		return nil, errors.New("start time cannot be after end time")
	}
	if breakMinutes < 0 {
		return nil, errors.New("break minutes cannot be negative")
	}

	return &Attendance{
		UserId:       userId,
		Date:         date,
		StartTime:    startTime,
		EndTime:      endTime,
		BreakMinutes: breakMinutes,
		Report:       report,
		CreatedAt:    time.Now(),
		UpdatedAt:    time.Now(),
	}, nil
}
func (a *Attendance) Validate() error {
	if a.UserId <= 0 {
		return errors.New("invalid user ID")
	}
	if a.StartTime.After(a.EndTime) {
		return errors.New("start time cannot be after end time")
	}
	if a.BreakMinutes < 0 {
		return errors.New("break minutes cannot be negative")
	}
	if a.Report == "" {
		return errors.New("report cannot be empty")
	}
	return nil
}
